### ISSUE TITLE
cf ssh -c command sometimes drops execution results

### ISSUE THREAD

**Issue by kei-yamazaki** [cloudfoundry-incubator/diego-ssh#11]
I tried ```cf ssh -c``` command 3 times but execution results is returned only 1 time.

```
$ cf ssh dora1 -c /usr/bin/env
$ cf ssh dora1 -c /usr/bin/env
$ cf ssh dora1 -c /usr/bin/env
CF_INSTANCE_ADDR=10.244.16.9:60000
TERM=screen-256color
USER=vcap
VCAP_APPLICATION={"limits":{"mem":256,"disk":1024,"fds":16384},"application_id":"baa4c41d-8b87-4589-bf56-c682e98220d9","application_version":"e1630967-2e74-45ed-bb23-cf5c06ee85c8","application_name":"dora1","application_uris":["dora1.bosh-lite.com"],"version":"e1630967-2e74-45ed-bb23-cf5c06ee85c8","name":"dora1","space_name":"kei-space","space_id":"ea7b18a2-accd-46f8-b989-13bc6b966b97","uris":["dora1.bosh-lite.com"]}
CF_INSTANCE_GUID=eb72a133-b168-42aa-76ba-214ee13981de
PATH=/bin:/usr/bin
PWD=/home/vcap
LANG=en_US.UTF-8
CF_INSTANCE_PORT=60000
VCAP_SERVICES={}
CF_INSTANCE_IP=10.244.16.9
CF_INSTANCE_INDEX=0
SHLVL=1
HOME=/home/vcap
CF_INSTANCE_PORTS=[{"external":60000,"internal":8080},{"external":60001,"internal":2222}]
INSTANCE_INDEX=0
PORT=8080
INSTANCE_GUID=eb72a133-b168-42aa-76ba-214ee13981de
MEMORY_LIMIT=256m
_=/usr/bin/env
```

Following is version of releases and CLI in problem occurs.

* CLI 6.14.0
* releases
  1. cf-release v225
  1. diego-release 0.1441.0
  1. garden-linux-release 0.327.0
  1. etcd-release 18

I deployed several different versions of release and CLI for separate problem.
But the problem occurs regardless version.

Next, I collected ssh-proxy's logs in each case.
When does not drop execution results, logs order is **"proxy-requests.request", "to-source.copy.completed"**.
But when drops execution results, logs order is **"to-source.copy.completed", "proxy-requests.request"**.

**When does not drop execution results**
```json
{"timestamp":"1448587770.403334379","source":"ssh-proxy","message":"ssh-proxy.authentication-failed","log_level":2,"data":{"error":"no auth passed yet","user":"cf:baa4c41d-8b87-4589-bf56-c682e98220d9/0"}}
{"timestamp":"1448587770.410322905","source":"ssh-proxy","message":"ssh-proxy.cf-authenticate.authenticate-starting","log_level":1,"data":{"session":"80"}}
{"timestamp":"1448587770.482597113","source":"ssh-proxy","message":"ssh-proxy.cf-authenticate.authenticate-finished","log_level":1,"data":{"session":"80"}}
{"timestamp":"1448587770.482841969","source":"ssh-proxy","message":"ssh-proxy.authentication-failed","log_level":2,"data":{"user":"cf:baa4c41d-8b87-4589-bf56-c682e98220d9/0"}}
{"timestamp":"1448587770.519655228","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.started","log_level":1,"data":{"session":"79.2"}}
{"timestamp":"1448587770.519894361","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.started","log_level":1,"data":{"session":"79.3"}}
{"timestamp":"1448587770.520037174","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.started","log_level":1,"data":{"session":"79.4"}}
{"timestamp":"1448587770.520130873","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.new-channel","log_level":1,"data":{"channelType":"session","extraData":"","session":"79.4"}}
{"timestamp":"1448587770.520245075","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.started","log_level":1,"data":{"session":"79.5"}}
{"timestamp":"1448587770.521113396","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-target.copy.started","log_level":1,"data":{"session":"79.4.1.1"}}
{"timestamp":"1448587770.521248579","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-source.copy.started","log_level":1,"data":{"session":"79.4.2.1"}}
{"timestamp":"1448587770.521396637","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.started","log_level":1,"data":{"channel-type":"session","session":"79.4.3"}}
{"timestamp":"1448587770.521519661","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.started","log_level":1,"data":{"channel-type":"session","session":"79.4.4"}}
{"timestamp":"1448587770.526620865","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.request","log_level":1,"data":{"channel-type":"session","payload":"AAAADC91c3IvYmluL2Vudg==","session":"79.4.3","type":"exec","wantReply":true}}
{"timestamp":"1448587770.529787302","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.request","log_level":1,"data":{"channel-type":"session","payload":"AAAAAA==","session":"79.4.4","type":"exit-status","wantReply":false}}
{"timestamp":"1448587770.530008554","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-source.copy.completed","log_level":1,"data":{"session":"79.4.2.1"}}
{"timestamp":"1448587770.530137300","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.completed","log_level":1,"data":{"channel-type":"session","session":"79.4.4"}}
{"timestamp":"1448587770.535077333","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-target.copy.completed","log_level":1,"data":{"session":"79.4.1.1"}}
{"timestamp":"1448587770.535240650","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.completed","log_level":1,"data":{"channel-type":"session","session":"79.4.3"}}
{"timestamp":"1448587770.537682533","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.completed","log_level":1,"data":{"session":"79.4"}}
{"timestamp":"1448587770.537858009","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.completed","log_level":1,"data":{"session":"79.2"}}
{"timestamp":"1448587770.537962198","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.completed","log_level":1,"data":{"session":"79.5"}}
{"timestamp":"1448587770.538049459","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.completed","log_level":1,"data":{"session":"79.3"}}
```

**When drops execution results**
```json
{"timestamp":"1448587806.308309317","source":"ssh-proxy","message":"ssh-proxy.authentication-failed","log_level":2,"data":{"error":"no auth passed yet","user":"cf:baa4c41d-8b87-4589-bf56-c682e98220d9/0"}}
{"timestamp":"1448587806.315051079","source":"ssh-proxy","message":"ssh-proxy.cf-authenticate.authenticate-starting","log_level":1,"data":{"session":"98"}}
{"timestamp":"1448587806.380640507","source":"ssh-proxy","message":"ssh-proxy.cf-authenticate.authenticate-finished","log_level":1,"data":{"session":"98"}}
{"timestamp":"1448587806.380898237","source":"ssh-proxy","message":"ssh-proxy.authentication-failed","log_level":2,"data":{"user":"cf:baa4c41d-8b87-4589-bf56-c682e98220d9/0"}}
{"timestamp":"1448587806.415349722","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.started","log_level":1,"data":{"session":"97.2"}}
{"timestamp":"1448587806.415685415","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.started","log_level":1,"data":{"session":"97.3"}}
{"timestamp":"1448587806.415838242","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.started","log_level":1,"data":{"session":"97.4"}}
{"timestamp":"1448587806.416002035","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.new-channel","log_level":1,"data":{"channelType":"session","extraData":"","session":"97.4"}}
{"timestamp":"1448587806.416140795","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.started","log_level":1,"data":{"session":"97.5"}}
{"timestamp":"1448587806.416851044","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-target.copy.started","log_level":1,"data":{"session":"97.4.1.1"}}
{"timestamp":"1448587806.417051792","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-source.copy.started","log_level":1,"data":{"session":"97.4.2.1"}}
{"timestamp":"1448587806.417159319","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.started","log_level":1,"data":{"channel-type":"session","session":"97.4.3"}}
{"timestamp":"1448587806.417259693","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.started","log_level":1,"data":{"channel-type":"session","session":"97.4.4"}}
{"timestamp":"1448587806.423390865","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.request","log_level":1,"data":{"channel-type":"session","payload":"AAAADC91c3IvYmluL2Vudg==","session":"97.4.3","type":"exec","wantReply":true}}
{"timestamp":"1448587806.426315069","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-source.copy.completed","log_level":1,"data":{"session":"97.4.2.1"}}
{"timestamp":"1448587806.426509380","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.request","log_level":1,"data":{"channel-type":"session","payload":"AAAAAA==","session":"97.4.4","type":"exit-status","wantReply":false}}
{"timestamp":"1448587806.426655769","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.completed","log_level":1,"data":{"channel-type":"session","session":"97.4.4"}}
{"timestamp":"1448587806.431985855","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.to-target.copy.completed","log_level":1,"data":{"session":"97.4.1.1"}}
{"timestamp":"1448587806.432124615","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.proxy-requests.completed","log_level":1,"data":{"channel-type":"session","session":"97.4.3"}}
{"timestamp":"1448587806.435985565","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.completed","log_level":1,"data":{"session":"97.4"}}
{"timestamp":"1448587806.436112881","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.completed","log_level":1,"data":{"session":"97.2"}}
{"timestamp":"1448587806.436230183","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-channels.completed","log_level":1,"data":{"session":"97.5"}}
{"timestamp":"1448587806.436335087","source":"ssh-proxy","message":"ssh-proxy.handle-connection.proxy-global-requests.completed","log_level":1,"data":{"session":"97.3"}}
```

So I found next code.

https://github.com/cloudfoundry-incubator/diego-ssh/blob/ecb7b1622db10af8fec8d8a48d927d99423f566b/proxy/proxy.go#L147-L157

This code seems to cause the thread race condition.

For experiment, added sleep code before ```ProxyRequests```, the problem occurs approximately 100%.

```go
go func() {
	helpers.Copy(logger.Session("to-target"), nil, targetChan, sourceChan)
	targetChan.CloseWrite()
}()
go func() {
	helpers.Copy(logger.Session("to-source"), nil, sourceChan, targetChan)
	sourceChan.CloseWrite()
}()

// Add this sleep, the problem occurs approximately 100%
time.Sleep(3*time.Second)

go ProxyRequests(logger, newChannel.ChannelType(), sourceReqs, targetChan)
go ProxyRequests(logger, newChannel.ChannelType(), targetReqs, sourceChan)
```

**Comment by ematpl** [comment]
Thanks for the thorough investigation and report, @kei-yamazaki! The Diego team will look into this issue shortly. 

Best,
Eric, CF Runtime Diego PM

**Comment by sykesm** [comment]
@emalm the problem is real but I don't think it's due to a race in the proxy. I added sleeps and rearranged the routines and did other crazy things without consistently recreating the issue.

I strongly suspect that the race is actually happening in the client side in the cli.  The streams are [copied in go routines][1] but we never [wait for the copies to complete][2].

It appears the behavior changed slightly with go 1.5 which is probably why you see the issue more often as the new cli is built with 1.5.

[1]: https://github.com/cloudfoundry/cli/blob/master/cf/ssh/ssh.go#L260-L261
[2]: https://github.com/cloudfoundry/cli/blob/master/cf/ssh/ssh.go#L289

**Comment by DanLavine** [comment]
Hi @kei-yamazaki,

This [PR](https://github.com/cloudfoundry-incubator/diego-ssh/pull/13) should address the issue once a new version of the CLI is released (the version as of writing this is 6.14.0). We are going to close this issue for now, but if this issue arises again on a later CLI version please feel free to open it back up.

Thanks,
@andrew-edgar && @danlavine, CF Diego Team

**Comment by ematpl** [comment]
To be fair, this should be fixed on the client side once https://github.com/cloudfoundry/cli/pull/702 is merged and a new CLI version released.

@kei-yamazaki, does this ever fail with a different client, such as OpenSSH? You should be able to try that with something like

```
ssh -p 2222 cf:$(cf app dora --guid)/0@ssh.bosh-lite.com /usr/bin/env
```

where as the password you supply an auth code given by `cf ssh-code`.

Thanks,
Eric

**Comment by kei-yamazaki** [comment]
@DanLavine @ematpl, Sorry slow response.

I tried new version ssh-proxy and CLI and this issue was resolved.

Following is the test results.

Result | Client | ssh-proxy
------------ | ------------ | ------------
NG | CLI 6.14.0 | v225
OK | OpenSSH 7.1p1 | v225
OK | CLI https://github.com/cloudfoundry/cli/commit/839f02b0138f3f52d768c80a9793d977667d9629 | https://github.com/cloudfoundry-incubator/diego-ssh/commit/18955d4a90fc1f96564e3909b2fbd45ac728d457 built with go1.4.2
OK | OpenSSH 7.1p1 | https://github.com/cloudfoundry-incubator/diego-ssh/commit/18955d4a90fc1f96564e3909b2fbd45ac728d457 built with go1.4.2
NG | CLI https://github.com/cloudfoundry/cli/commit/839f02b0138f3f52d768c80a9793d977667d9629 | https://github.com/cloudfoundry-incubator/diego-ssh/commit/18955d4a90fc1f96564e3909b2fbd45ac728d457 built with go1.5.1
NG | OpenSSH 7.1p1 | https://github.com/cloudfoundry-incubator/diego-ssh/commit/18955d4a90fc1f96564e3909b2fbd45ac728d457 built with go1.5.1

Thanks @sykesm

**Comment by ematpl** [comment]
Thanks for the thorough report, @kei-yamazaki! We do plan to move diego-release to Go 1.5 eventually, so we'll have to watch out for that behavior in the ssh-proxy when we do.

Best,
Eric

**Comment by dkoper** [comment]
FYI, we have merged PR#702 to the CLI repo. Please try it out with our edge binary and let us know.

Cheers,
Dies Koper
CF CLI PM

**Comment by kei-yamazaki** [comment]
@dkoper, Thank you for your information

Already tried the edge binary, please refer [previous comment](https://github.com/cloudfoundry-incubator/diego-ssh/issues/11#issuecomment-162769427) (cloudfoundry/cli@839f02b is it).
